File: test-fixtures/security/vulnerable.ts
// INTENTIONAL VULNERABILITIES - DO NOT USE IN PRODUCTION
// Test fixture for security scanner validation

// TS-001: Any type usage (bypasses type safety)
function processData(data: any): any { // BAD: any parameter
    const result: any = data.transform(); // BAD: any variable
    return result as any; // BAD: cast to any
}

interface UnsafeConfig {
    options: any; // BAD: any in interface
    handler: <any>(x: any) => any; // BAD: generic any
}

// TS-002: Non-null assertion operator (can cause runtime errors)
function getUserName(user?: User): string {
    return user!.name; // BAD: Non-null assertion on optional
}

function processItems(items?: Item[]): void {
    const first = items![0]; // BAD: Non-null assertion on optional array
    const length = items!.length; // BAD
}

class UserService {
    private user?: User;

    getName(): string {
        return this.user!.name; // BAD: Non-null assertion
    }

    getItems(): Item[] {
        return this.user!.items; // BAD: Non-null assertion
    }
}

// SAFE PATTERNS (should not trigger)
function safeProcess(data: unknown): string {
    if (typeof data === 'string') {
        return data; // GOOD: Type narrowing
    }
    return 'default';
}

function safeGetName(user?: User): string {
    return user?.name ?? 'Anonymous'; // GOOD: Optional chaining
}

function safeGetItems(items?: Item[]): Item | undefined {
    return items?.[0]; // GOOD: Optional chaining
}

// Types for reference
interface User {
    name: string;
    items: Item[];
}

interface Item {
    id: number;
    value: string;
}

File: test-fixtures/security/vulnerable.rs
// INTENTIONAL VULNERABILITIES - DO NOT USE IN PRODUCTION
// Test fixture for security scanner validation

// RUST-001: Unsafe blocks without proper documentation
fn process_data(ptr: *const u8, len: usize) {
    unsafe {  // BAD: No SAFETY comment
        let slice = std::slice::from_raw_parts(ptr, len);
        println!("{:?}", slice);
    }
}

// RUST-002: Unchecked unwrap on fallible operations
fn parse_config(data: &str) -> i32 {
    let value = data.parse::<i32>().unwrap();  // BAD: May panic
    let other = Some(42).expect("hardcoded");   // BAD: May panic
    value + other
}

// RUST-003: Raw pointer operations
fn raw_pointer_usage() {
    let x = 42;
    let ptr = &x as *const i32;           // BAD: Raw pointer cast
    let raw = x.to_string().as_ptr();     // BAD: as_ptr() call
    unsafe {
        std::ptr::read(ptr);               // BAD: ptr::read
        std::ptr::write(ptr as *mut i32, 0); // BAD: ptr::write
    }
}

// SAFE PATTERNS (should not trigger or have reduced severity)
fn safe_code_example() {
    // SAFETY: ptr is guaranteed valid by caller contract
    unsafe {
        // documented unsafe block
    }

    // Using proper error handling
    let value = "42".parse::<i32>().unwrap_or(0);
    let other = Some(42).unwrap_or_default();
    let result = "123".parse::<i32>()?;  // Would need Result return type

    // Using NonNull for safer pointer handling
    let x = 42;
    if let Some(ptr) = std::ptr::NonNull::new(&x as *const _ as *mut i32) {
        // Safe: using NonNull abstraction
    }
}
